src/native/eventpipe/ds-ipc-pal-socket.c

```diff
@@ -0,0 +1,1561 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#ifdef FEATURE_PERFTRACING_STANDALONE_PAL
+#define EP_NO_RT_DEPENDENCY
+#endif
+
+#include "ds-rt-config.h"
+
+#ifdef ENABLE_PERFTRACING
+
+#define DS_IMPL_IPC_PAL_SOCKET_GETTER_SETTER
+#include "ds-ipc-pal-socket.h"
+
+#if defined (DS_IPC_PAL_TCP)
+#define DS_IPC_PAL_AF_INET
+#define DS_IPC_PAL_AF_INET6
+#elif defined (DS_IPC_PAL_UDS)
+#define DS_IPC_PAL_AF_UNIX
+#else
+#error "Unsupported PAL socket configuration"
+#endif
+
+#ifndef HOST_WIN32
+#include <assert.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <unistd.h>
+#include <fcntl.h>
+#include <errno.h>
+#include <sys/un.h>
+#include <sys/stat.h>
+
+#if __GNUC__
+#include <poll.h>
+#else
+#include <sys/poll.h>
+#endif
+
+#ifdef DS_IPC_PAL_TCP
+#include <netinet/in.h>
+#include <netinet/tcp.h>
+#include <netdb.h>
+#endif
+#endif
+
+#ifdef HOST_WIN32
+#define DS_IPC_INVALID_SOCKET INVALID_SOCKET
+#define DS_IPC_SOCKET_ERROR SOCKET_ERROR
+#define DS_IPC_SOCKET_ERROR_WOULDBLOCK WSAEWOULDBLOCK
+#define DS_IPC_SOCKET_ERROR_TIMEDOUT WSAETIMEDOUT
+#define DS_IPC_ERROR_INTERRUPT WSAEINTR
+typedef ADDRINFOA ds_ipc_addrinfo_t;
+typedef WSAPOLLFD ds_ipc_pollfd_t;
+typedef int ds_ipc_mode_t;
+#else
+#define DS_IPC_INVALID_SOCKET -1
+#define DS_IPC_SOCKET_ERROR -1
+#define DS_IPC_SOCKET_ERROR_WOULDBLOCK EINPROGRESS
+#define DS_IPC_SOCKET_ERROR_TIMEDOUT ETIMEDOUT
+#define DS_IPC_ERROR_INTERRUPT EINTR
+typedef struct addrinfo ds_ipc_addrinfo_t;
+typedef struct pollfd ds_ipc_pollfd_t;
+typedef mode_t ds_ipc_mode_t;
+#endif
+
+#ifndef EP_NO_RT_DEPENDENCY
+#include "ds-rt.h"
+#else
+#ifdef FEATURE_CORECLR
+#include <pal.h>
+#include "processdescriptor.h"
+#endif
+
+#ifndef ep_return_null_if_nok
+#define ep_return_null_if_nok(expr) do { if (!(expr)) return NULL; } while (0)
+#endif
+
+#ifndef ep_raise_error_if_nok
+#define ep_raise_error_if_nok(expr) do { if (!(expr)) goto ep_on_error; } while (0)
+#endif
+
+#ifndef ep_raise_error
+#define ep_raise_error() do { goto ep_on_error; } while (0)
+#endif
+
+#ifndef ep_exit_error_handler
+#define ep_exit_error_handler() do { goto ep_on_exit; } while (0)
+#endif
+
+#ifndef EP_ASSERT
+#define EP_ASSERT assert
+#endif
+
+#ifndef DS_ENTER_BLOCKING_PAL_SECTION
+#define DS_ENTER_BLOCKING_PAL_SECTION
+#endif
```

Comment:
No, we can have dependency on EventPipe library from Diagnostic Server (but not the other way around). A couple needs to be redefined like this since we can build PAL standalone in debug_pal, meaning we can't bring in the full definitions from EventPipe since that also brings in the runtime shim implementation and that can't be done when building debug_pal.